### ISSUE TITLE
Scitoken.conf

### ISSUE THREAD

**Issue by biozit** [xrootd/xrootd#2074]
Hi All,

   Using this config for scitokens, the Xrootd says the latest issuer is not listed:

`[Global]
audience = IceCube, NRP

[Issuer https://chtc.cs.wisc.edu/icecube]
issuer = https://chtc.cs.wisc.edu/icecube
base_path = /icecube
map_subject = False

[Issuer https://token.nationalresearchplatform.org/modis]
issuer = https://token.nationalresearchplatform.org/modis
base_path = /nrp/protected/sio/MODIS_Aqua_microphysics_images/
map_subject = False

[Issuer https://token.nationalresearchplatform.org/ucsdphysics]
issuer = https://token.nationalresearchplatform.org/ucsdphysics
base_path = /ucsd/physics
map_subject = False`

However, XrootD is ok with this:

`[Global]
audience = IceCube, NRP

[Issuer https://chtc.cs.wisc.edu/icecube]
issuer = https://chtc.cs.wisc.edu/icecube
base_path = /icecube
map_subject = False

[Issuer https://token.nationalresearchplatform.org/modis]
issuer = https://token.nationalresearchplatform.org/modis
base_path = /nrp/protected/sio/MODIS_Aqua_microphysics_images/
map_subject = False

[Issuer https://**token.nationalresearchplatform.ort**/ucsdphysics]
issuer = https://**token.nationalresearchplatform.ort**/ucsdphysics
base_path = /ucsd/physics
map_subject = False
`

**Comment by abh3** [comment]
Hmm, which issues would be the latest? Can you turn on some debugging to 
see what xrootd (well actually the SciTokens plugin) thinks is going on?

**Comment by biozit** [comment]
`stash-origin-auth]# xrootd -version
v5.5.5`

This is the case with 2 issuers:

`[root@node-1-5 stash-origin-auth]# cat scitokens.conf 
[Global]
audience = IceCube, NRP

[Issuer https://chtc.cs.wisc.edu/icecube]
issuer = https://chtc.cs.wisc.edu/icecube
base_path = /icecube
map_subject = False

[Issuer https://token.nationalresearchplatform.org/modis]
issuer = https://token.nationalresearchplatform.org/modis
base_path = /nrp/protected/sio/MODIS_Aqua_microphysics_images/
map_subject = False


[Issuer https://token.nationalresearchplatform.org/ucsdphysics]
issuer = https://token.nationalresearchplatform.org/ucsdphysics
base_path = /ucsd/physics
map_subject = False`

`Plugin loaded XRootDTCP v5.4.0 from xrd.tcpmonlib /usr/lib64/libXrdTCPStats-5.so
------ xrootd stash-origin-auth@sdsc-origin.nationalresearchplatform.org:1095 initialization completed.
230822 18:24:11 291323 XrootdBridge: unknown.1:35@ip68-8-122-198.sd.sd.cox.net login as nobody
230822 18:24:11 291323 scitokens_GenerateAcls: Failed to deserialize SciToken: token verification failed: Token issuer is not in list of allowed issuers.
230822 18:24:11 291323 scitokens_Access: Failed to generate ACLs for token
230822 18:24:11 291323 acc_Audit: unknown.1:35@ip68-8-122-198.sd.sd.cox.net deny https *@[::ffff:68.8.122.198] create /nrp/protected/sio/MODIS_Aqua_microphysics_images/aaa
230822 18:24:11 291323 scitokens_GenerateAcls: Failed to deserialize SciToken: token verification failed: Token issuer is not in list of allowed issuers.
230822 18:24:11 291323 scitokens_Access: Failed to generate ACLs for token
230822 18:24:11 291323 acc_Audit: unknown.1:35@ip68-8-122-198.sd.sd.cox.net deny https *@[::ffff:68.8.122.198] excl_create /nrp/protected/sio/MODIS_Aqua_microphysics_images/aaa
230822 18:24:11 291323 ofs_open: unknown.1:35@ip68-8-122-198.sd.sd.cox.net Unable to create /nrp/protected/sio/MODIS_Aqua_microphysics_images/aaa; permission denied
230822 18:24:11 291323 unknown.1:35@ip68-8-122-198.sd.sd.cox.net Xrootd_Response: sending err 3010: Unable to create /nrp/protected/sio/MODIS_Aqua_microphysics_images/aaa; permission denied
230822 18:24:11 291323 XrootdXeq: unknown.1:35@ip68-8-122-198.sd.sd.cox.net disc 0:00:00 (send failure)
230822 18:24:15 291322 XrootdBridge: unknown.2:36@ip68-8-122-198.sd.sd.cox.net login as nobody
230822 18:24:15 291322 scitokens_GenerateAcls: Failed to deserialize SciToken: token verification failed: Token issuer is not in list of allowed issuers.
230822 18:24:15 291322 scitokens_Access: Failed to generate ACLs for token
230822 18:24:15 291322 acc_Audit: unknown.2:36@ip68-8-122-198.sd.sd.cox.net deny https *@[::ffff:68.8.122.198] create /nrp/protected/sio/MODIS_Aqua_microphysics_images/aaa
230822 18:24:15 291322 scitokens_GenerateAcls: Failed to deserialize SciToken: token verification failed: Token issuer is not in list of allowed issuers.
230822 18:24:15 291322 scitokens_Access: Failed to generate ACLs for token
230822 18:24:15 291322 acc_Audit: unknown.2:36@ip68-8-122-198.sd.sd.cox.net deny https *@[::ffff:68.8.122.198] excl_create /nrp/protected/sio/MODIS_Aqua_microphysics_images/aaa
`

This case is with this config, two different domains:

`[Global]
audience = IceCube, NRP

[Issuer https://chtc.cs.wisc.edu/icecube]
issuer = https://chtc.cs.wisc.edu/icecube
base_path = /icecube
map_subject = False

[Issuer https://token.nationalresearchplatform.org/modis]
issuer = https://token.nationalresearchplatform.org/modis
base_path = /nrp/protected/sio/MODIS_Aqua_microphysics_images/
map_subject = False


[Issuer https://token.nationalresearchplatform.ort/ucsdphysics]
issuer = https://token.nationalresearchplatform.ort/ucsdphysics
base_path = /ucsd/physics
map_subject = False`

`230822 18:30:15 292637 scitokens_Access: Trying token-based access control
230822 18:30:15 292637 scitokens_Access: Cached token mapped_username=, subject=NRP, issuer=https://token.nationalresearchplatform.org/modis, authorizations=/nrp/protected/sio/MODIS_Aqua_microphysics_images:read,dir,stat,create,mkdir,mv,insert,update,chmod,del
230822 18:30:15 292637 scitokens_Access: Grant authorization based on scopes for operation=create, path=/nrp/protected/sio/MODIS_Aqua_microphysics_images/aaa
230822 18:30:15 292637 unknown.2:37@ip68-8-122-198.sd.sd.cox.net ofs_fstat:  fn=/nrp/protected/sio/MODIS_Aqua_microphysics_images/aaa
`

**Comment by olifre** [comment]
Maybe this is due to `inih`'s limitation of the section title to 50 bytes, i.e. 49 characters plus trailing `\0`?
https://github.com/xrootd/xrootd/blob/8f0a3ebf6b14c65c57667d07c0c624f3d4b215a8/src/XrdSciTokens/vendor/inih/INIReader.h#L120

This matches:
```
echo -n "Issuer https://token.nationalresearchplatform.org" | wc -c
49
```
i.e the section titles must differ within the first 49 characters, otherwise the last section "wins" and overrides the earlier one.

**Comment by biozit** [comment]
Yes, it looks like the case. This is my scitoken config file, and it is working:

[Global]
audience = IceCube, NRP

[Issuer https://chtc.cs.wisc.edu/icecube]
issuer = https://chtc.cs.wisc.edu/icecube
base_path = /icecube
map_subject = False

[Issuer https://token.nationalresearchplatform.org/modis]
issuer = https://token.nationalresearchplatform.org/modis
base_path = /nrp/protected/sio/MODIS_Aqua_microphysics_images/
map_subject = False

# The following lines are from scitokens.conf.local:
[Issuer https://t.nationalresearchplatform.org/ucsdphysics]
issuer = https://t.nationalresearchplatform.org/ucsdphysics
base_path = /ucsd/physics
map_subject = False

**Comment by abh3** [comment]
Let's discuss this issue at the Xcache meeting Thursday.

**Comment by amadio** [comment]
Fixed by #2079.
